File: lab-grpc/java/grpc-central/src/main/java/br/pucminas/labdamd/central/ClienteCentral.java
package br.pucminas.labdamd.central;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.Scanner;

// Parte C/D - cliente gRPC. O que dá pra reparar: eu chamo métodos do stub como
// se fossem funções locais. Não tem send/receive nem montagem de string - o
// gRPC serializa, manda por HTTP/2 e desserializa a resposta por baixo.
public class ClienteCentral {
    // OFFSET = mesmo valor do servidor (ver README)
    static final int OFFSET = 39;

    public static void main(String[] args) {
        int porta = 50051 + OFFSET; // 50090
        // O "canal" é a conexão com o servidor. usePlaintext() = sem TLS (lab local).
        ManagedChannel canal = ManagedChannelBuilder.forAddress("localhost", porta)
                .usePlaintext()
                .build();
        try {
            CentralAtendimentoGrpc.CentralAtendimentoBlockingStub stub =
                    CentralAtendimentoGrpc.newBlockingStub(canal);

            Scanner teclado = new Scanner(System.in);
            System.out.print("Digite seu nome: ");
            String nome = teclado.nextLine();

            // Chamada unária: parece uma chamada de método local, mas atravessa a rede.
            PerguntaHorario pergunta = PerguntaHorario.newBuilder().setNomeAluno(nome).build();
            RespostaHorario resposta = stub.consultarHorario(pergunta);
            System.out.println("[gRPC] " + resposta.getMensagem());

            // Chamada com streaming: uma inscrição, vários Avisos ao longo do tempo.
            // O stub devolve um Iterator; cada next() é um aviso que o servidor mandou.
            System.out.println("[gRPC] Inscrevendo-se para acompanhar avisos...");
            InscricaoAvisos inscricao = InscricaoAvisos.newBuilder().setNomeAluno(nome).build();
            java.util.Iterator<Aviso> avisos = stub.acompanharAvisos(inscricao);
            while (avisos.hasNext()) {
                Aviso aviso = avisos.next();
                System.out.println("[gRPC] Recebido: " + aviso.getTexto());
            }
        } finally {
            canal.shutdown();
        }
    }
}

File: lab-grpc/python/grpc_central/cliente_central.py
import grpc
import central_pb2
import central_pb2_grpc

# Parte C/D - cliente gRPC em Python. Chamo o método do stub como se fosse uma
# função local; o gRPC serializa, manda pela rede e desserializa a resposta.
# OFFSET = mesmo valor do servidor (ver README)
OFFSET = 39
PORTA = 50061 + OFFSET  # 50100


def main():
    canal = grpc.insecure_channel(f"localhost:{PORTA}")
    stub = central_pb2_grpc.CentralAtendimentoStub(canal)

    nome = input("Digite seu nome: ")

    # Chamada unária: parece uma chamada de função local, mas atravessa a rede.
    resposta = stub.ConsultarHorario(central_pb2.PerguntaHorario(nome_aluno=nome))
    print(f"[gRPC] {resposta.mensagem}")


if __name__ == "__main__":
    main()
